### ISSUE TITLE
Connect to flask-jsonrpc

### ISSUE THREAD

**Issue by ne0h** [cinemast/libjson-rpc-cpp#119]
Hi,

we are trying to build a simple chat service (https://github.com/ne0h/catmail) that uses json rpc.
While the client is implemented in c++, the server uses python with flask-jsonrpc.

When we try to send a simple query to the server we get this error:
`libc++abi.dylib: terminating with unexpected exception of type Json::LogicError: Value is not convertible to Int.`

You are free to access our test service at `http://miauuu.de:5000` and try the query by yourself.

Thanks in advance,
Max

**Comment by ne0h** [comment]
You can use curl to send the query:
`curl -i -X POST -d '{"jsonrpc": "2.0", "method": "createUser", "params": {"username" : "cat", "password" : "fewfwfwfwef"}, "id": "12343244"}' http://miauuu.de:5000/`

The server returns:
`HTTP/1.0 200 OK
Content-Type: application/json
Content-Length: 79
Server: Werkzeug/0.10.4 Python/2.7.3
Date: Wed, 22 Jul 2015 14:23:03 GMT

{
  "id": "12343244", 
  "jsonrpc": "2.0", 
  "result": {
    "result": 1
  }`

**Comment by ne0h** [comment]
This is the return output of the server:
`HTTP/1.0 200 OK
Content-Type: application/json
Content-Length: 53
Server: Werkzeug/0.10.4 Python/2.7.3
Date: Wed, 22 Jul 2015 15:26:52 GMT

{
  "id": "1", 
  "jsonrpc": "2.0", 
  "result": -1
}`

Does libjson-rpc-cpp not like whitespaces or line breaks?

**Comment by cinemast** [comment]
I will take a look at it within the next days.

**Comment by cinemast** [comment]
Do you also have snippets how you use libjson-rpc-cpp?

**Comment by ne0h** [comment]
Thanks a lot! Hier is the specific code. The error occurs before the debug output in line 14.

**Comment by cinemast** [comment]
Where is it? :D

**Comment by ne0h** [comment]
Oh i am sorry, i forgot to paste the link :D
https://github.com/ne0h/catmail/blob/master/client/src/client.cpp

**Comment by cinemast** [comment]
You are using the "id" field as int. The json-rpc standard requires the id field to be a string. Therefore you get the above exception. 

I have to ask: Why are you not using the jsonrpcstub tool, which generates the client for you? This could probably have avoided such errors. It would be interesting for me, why the tools was not usable in your scenario.

Greetings
Peter

**Comment by cinemast** [comment]
Also please close the issue, if everything is clear for you.

**Comment by ne0h** [comment]
I'm have updated the code since yesterday, but i still have the same error.

In `https://github.com/ne0h/catmail/blob/master/client/genapi.sh` we already use the jsonrpcstub tool to generate the interface. Is `json.spec` correct?

Thanks a lot!
Max

**Comment by cinemast** [comment]
No it is not correct. You don't have to bother about the json-rpc stuff. The framework does it all for you. All you have to define is which parameters your json-rpc method expect and what type they return.

So the correct spec would be: 

```json
[
	{
		"name": "createUser",
		"params": {
			"username": "cat",
			"password": "password"
		},
		"returns": 0
	}
]
```

**Comment by cinemast** [comment]
Also take a look at the example description: https://github.com/cinemast/libjson-rpc-cpp#step-1-writing-the-specification-file

**Comment by ne0h** [comment]
Of course i used this snippet as a start. I have updated `spec.json` but still the same error.

**Comment by cinemast** [comment]
Can you start a debugger and figure out which line the exception throws? It is odd that this can occur.

**Comment by ne0h** [comment]
It took me a while to build libjson-rpc-cpp and json-cpp as submodules with debug symboles. Here is the result, i do not know how to get it better:

`➜  build git:(master) ✗ gdb ./catmail-client 
GNU gdb (Gentoo 7.7.1 p1) 7.7.1
Copyright (C) 2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://bugs.gentoo.org/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./catmail-client...done.
(gdb) run
Starting program: /media/lds/Projekte/catmail/client/build/catmail-client 
warning: Could not load shared library symbols for linux-vdso.so.1.
Do you need "set solib-search-path" or "set sysroot"?
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
terminate called after throwing an instance of 'Json::LogicError'
  what():  Value is not convertible to Int.

Program received signal SIGABRT, Aborted.
0x00007ffff54e1774 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:55
55	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.`

**Comment by cinemast** [comment]
Please also add the backtrace (by typing `bt` after run). This information is not very helpful yet.

**Comment by ne0h** [comment]
I'm sorry. This is my first c++ project and i am not so firm with the tools so far. I hope this is more helpful:

(gdb) bt
#0  0x00007ffff54e1774 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:55
#1  0x00007ffff54e2929 in __GI_abort () at abort.c:89
#2  0x00007ffff5db2505 in __gnu_cxx::__verbose_terminate_handler() ()
   from /usr/lib/gcc/x86_64-pc-linux-gnu/4.8.4/libstdc++.so.6
#3  0x00007ffff5db06a6 in ?? () from /usr/lib/gcc/x86_64-pc-linux-gnu/4.8.4/libstdc++.so.6
#4  0x00007ffff5db06d3 in std::terminate() ()
   from /usr/lib/gcc/x86_64-pc-linux-gnu/4.8.4/libstdc++.so.6
#5  0x00007ffff5db06e6 in ?? () from /usr/lib/gcc/x86_64-pc-linux-gnu/4.8.4/libstdc++.so.6
#6  0x00007ffff5db0362 in __cxa_call_unexpected ()
   from /usr/lib/gcc/x86_64-pc-linux-gnu/4.8.4/libstdc++.so.6
#7  0x00007ffff707f6d3 in jsonrpc::RpcProtocolClient::HandleResponse (this=0x6846c0, value=..., 
    result=...)
    at /home/ne0h/Projekte/catmail/3rdparty/libjson-rpc-cpp/src/jsonrpccpp/client/rpcprotocolclient.cpp:53
#8  0x00007ffff707f018 in jsonrpc::RpcProtocolClient::HandleResponse (this=0x6846c0, 
    response="{\n  \"id\": \"1\", \n  \"jsonrpc\": \"2.0\", \n  \"result\": -1\n}", result=...)
    at /home/ne0h/Projekte/catmail/3rdparty/libjson-rpc-cpp/src/jsonrpccpp/client/rpcprotocolclient.cpp:45
#9  0x00007ffff707ab68 in jsonrpc::Client::CallMethod (this=0x684820, name="createUser", 
    parameter=..., result=...)
    at /home/ne0h/Projekte/catmail/3rdparty/libjson-rpc-cpp/src/jsonrpccpp/client/client.cpp:31
#10 0x00007ffff707be7c in jsonrpc::Client::CallMethod (this=0x684820, name="createUser", 
    parameter=...)
    at /home/ne0h/Projekte/catmail/3rdparty/libjson-rpc-cpp/src/jsonrpccpp/client/client.cpp:77
#11 0x00000000004152aa in ClientHandler::createUser (this=0x684820, password="cat_key", 
    username="cat") at /home/ne0h/Projekte/catmail/client/src/../include/../api/clienthandler.hpp:20
#12 0x0000000000414e82 in Client::createUser (this=0x6847c0, username="cat", password="cat_key", 
    userKeyPair=std::shared_ptr (count 2, weak 0) 0x684710, 
    exchangeKeyPair=std::shared_ptr (count 2, weak 0) 0x6855d0)
    at /home/ne0h/Projekte/catmail/client/src/client.cpp:13
#13 0x0000000000412b01 in main (argc=1, argv=0x7fffffffdf08)
    at /home/ne0h/Projekte/catmail/client/src/main.cpp:23

**Comment by cinemast** [comment]
Thank you for providing the backtrace. Don't be sorry :)

Well actually I think I detected two errors, one is on my side. libjson-rpc-cpp client always assumes that IDs are convertible to int. The standard allows it to be int, string or null. So I need to fix that.

On the other hand flask-jsonrpc is also not behaving correctly. If you send an int id field, you should also receive back an int id field in the response. Flask-jsonrpc always translates the id field to a string, regardless of what it was.

See here: 

Call with string -> return as string
```
curl -i -X POST -d '{"jsonrpc": "2.0", "method": "createUser", "params": {"username" : "cat", "password" : "fewfwfwfwef"}, "id": "12343244"}' http://miauuu.de:5000/
HTTP/1.0 200 OK
Content-Type: application/json
Content-Length: 60
Server: Werkzeug/0.10.4 Python/2.7.3
Date: Fri, 24 Jul 2015 10:50:45 GMT

{
  "id": "12343244", 
  "jsonrpc": "2.0", 
  "result": -1
} 
```                                                                                                                                                          
Call with int -> Response as string                                                                                                                                                  

```                                                                                                                                         
 curl -i -X POST -d '{"jsonrpc": "2.0", "method": "createUser", "params": {"username" : "cat", "password" : "fewfwfwfwef"}, "id": 12343244}' http://miauuu.de:5000/  
HTTP/1.0 200 OK
Content-Type: application/json
Content-Length: 60
Server: Werkzeug/0.10.4 Python/2.7.3
Date: Fri, 24 Jul 2015 10:50:51 GMT

{
  "id": "12343244", 
  "jsonrpc": "2.0", 
  "result": -1
} 
```

Do you see what I mean?

Thank you for taking the time reporting this. Fixing this error on my side might take a while, because it messes up the ABI and API compatibility. I will fix it first on the develop branch. Would this be ok for you?

**Comment by ne0h** [comment]
Thanks a lot for your work! We would be really happy using libjson-rpc-cpp in our chat. Development branch would be absolutely okay. That is why I included ibjson-rpc-cpp as a submodule, so I can switch to another branch.
Do you already know when you are having the time to fix it?

**Comment by cinemast** [comment]
I think it should be fixed for now :) Could you please validate this?

**Comment by ne0h** [comment]
Yes, it works. Thanks a lot for your work and for your patience!

**Comment by cinemast** [comment]
You're welcome. I think you should also report this to flask-jsonrpc, because its implementation is violating the json-rpc specification. Would you mind doing that?

**Comment by cinemast** [comment]
Is this the project you are using on the server side? https://github.com/cenobites/flask-jsonrpc

**Comment by ne0h** [comment]
Of course I will do this. I'm not sure if this is the correct flask project, I going to ask my team members who code the server.

**Comment by ne0h** [comment]
Will be fixed in 0.3.

**Comment by cinemast** [comment]
Thanks. :+1:
